torch/optim/lr_scheduler.py

```diff
@@ -26,6 +43,7 @@ def __init__(self, optimizer, last_epoch=-1):
         self.base_lrs = list(map(lambda group: group['initial_lr'], optimizer.param_groups))
         self.last_epoch = last_epoch
```

Comment:
I'm assuming the `-1` value is a leftover from when shedulers were called at the beginning of their epoch, and so the previous epoch of 0 would have been -1. Since we are now stepping schedulers at the end of the iteration, it is indeed confusing. It would make sense to default to `last_epoch=0` instead.

Reply:
> I think there should be a clear description of what `last_epoch == -1` means, it would just be nice if it were spelled out explicitly.

`last_epoch` is a scheduler-specific concept, and many schedulers in chainable form do not need an explicit `last_epoch`. The concept is supposed to say "This is the epoch that was last invoked. Before the scheduler's creation, `last_epoch` is -1, immediately after creation `last_epoch` is then 0, etc.

In fact, we should also deprecate `last_epoch` from `_LRScheduler`'s initialization. This means that other schedulers will need to re-implement that logic per-scheduler.

Thoughts?

> This is especially because previously we set `last_epoch == -1` to `0`, and now we don't anymore.

I consider that setting `last_epoch` to 0 before was a bug, see [here](https://github.com/pytorch/pytorch/pull/23306#discussion_r307434811) for our previous discussion on the meaning of the change.

Reply:
For reference: the impact of `last_epoch=-1` is discussed in #7889.